flink-ml-iteration/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/DataCacheWriter.java

```diff
@@ -43,6 +43,10 @@
     private final List<Segment> finishSegments;
 
     private SegmentWriter currentSegment;
+    /**
+     * The segments that are newly added that has not been retrieved by getNewlyFinishedSegments().
+     */
+    private final List<Segment> newlyFinishedSegments;
```

Comment:
Do we still need `newlyFinishedSegments` if on initialization we could add also include the pending segments into the `finishSegments` of the new writers?

Reply:
Hi Yun, this is a good point. But I think maintaining `newlyFinishedSegments` is better. The reason is as follows:

We need to maintain a not-yet processed segment list for each input (i.e., `segmentLists[inputIndex]`), i.e., in this list we need to (1) add the newly cached segments and (2) remove the processed segments.

If we are not maintaining the not-yet processed segment list and include the pending segments in  `finishSegmets` of the new writers, then we are not able to remove the processed segments.